### PR TITLE
Fixed layer_names.hpp

### PR DESCRIPTION
**Pull request by prince776**
Hi, I've added #ifndef so that layer_names.hpp can be included safely in multiple files.
I've also noticed some layers' names are not added in this file. Should I add them also in this PR?

### CODE REVIEW

**Comment by zoq** [comment]
If you like to add the new layer names part of the PR, please feel free.

File: src/mlpack/methods/ann/layer_names.hpp [hunk 1]
```diff
@@ -372,7 +440,238 @@ class LayerNameVisitor : public boost::static_visitor<std::string>
...
+    return "swish";
+  }
+
+  /**
+   * Return the name of the given layer of type MishFunction as a string.
+   * 
+   * @param Given layer of type MishFunction.
+   * @return The string representation of the layer.
+   */
+  std::string LayerString(MishFunction<>* /*layer*/) const
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer_names.hpp]: Hmm, This should be MishFunctionLayer

File: src/mlpack/methods/ann/layer_names.hpp [hunk 2]
```diff
@@ -372,7 +440,238 @@ class LayerNameVisitor : public boost::static_visitor<std::string>
...
+    return "mish";
+  }
+
+  /**
+   * Return the name of the given layer of type LiSHTFunction as a string.
+   * 
+   * @param Given layer of type LiSHTFunction.
+   * @return The string representation of the layer.
+   */
+  std::string LayerString(LiSHTFunction<>* /*layer*/) const
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer_names.hpp]: This should be LiSHTFunctionLayer.

**Comment by zoq** [comment]
Can you adjust the test suite as well?

**Comment by sreenikSS** [comment]
Looks good to me too. Just the tests and we are done.

However, @prince776 don't you think that as new layers get added one might miss out on modifying this file. What do you think about a test that will fail if any layer is present but is not defined here?
Even if not all but the ones that are of type ```LayerTypes```.

That being said, it would be necessary to ensure that all existing layers are already added here so that the test does not fail immediately.

@zoq I think previously this file was previously being implemented using visitors and there it was automatically not compiling until all the combinations were given. Can we take a cue from that if we decide to implement this?

**Comment by prince776** [comment]
> Looks good to me too. Just the tests and we are done.
> 
> However, @prince776 don't you think that as new layers get added one might miss out on modifying this file. What do you think about a test that will fail if any layer is present but is not defined here?
> Even if not all but the ones that are of type `LayerTypes`.
> 
> That being said, it would be necessary to ensure that all existing layers are already added here so that the test does not fail immediately.
> 
> @zoq I think previously this file was previously being implemented using visitors and there it was automatically not compiling until all the combinations were given. Can we take a cue from that if we decide to implement this?

Hi @sreenikSS, I was also thinking the same, since when I wrote the tests it seemed really redundant. I was getting some error due to LayerTypes so it was taking a little too much time(sometimes I get stuck of really trivial errors). 
I was thinking of proposing the idea to add a string member to layer class that will be returned by layer_name visitor or we can have the layer_name visitor return the name of class converted to string. What are your thoughts about this?

**Comment by prince776** [comment]
Hi, I've added test cases for all Layers except `unsupported` because to add that we need a layer that is not in `LayerTypes` but when we do that, it'll give an incomprehensible error during compilation.

Also there is possibility of custom layers to be added, then that will return 'unsupported' however that is misleading. So I think each layer should have a member `std::string name` which can be returned.

**Comment by sreenikSS** [comment]
> Hi, I've added test cases for all Layers except unsupported because to add that we need a layer that is not in LayerTypes but when we do that, it'll give an incomprehensible error during compilation.

That is understandable since all possible layer types are supposed to be included here. However, what happens when you pass an instance of type ```LayerTypes<>```?

> Also there is possibility of custom layers to be added, then that will return 'unsupported' however that is misleading.

You mean to create a dummy layer, right? That should work too but try with ```LayerTypes<>``` once before doing that.

> So I think each layer should have a member std::string name which can be returned.

This is a valid point. I think there has been a debate regarding this thing before also. In fact if I remember correctly the layers did have a ```string name``` parameter that could be returned (this was a few years back) but was eventually removed due to maintenance issues (not too sure if I remember the reason correctly).
Since this is a design issue, I guess @rcurtin and @zoq would be able to throw more light on why exactly the parameter was discarded.

**Comment by sreenikSS** [comment]
> we can have the layer_name visitor return the name of class converted to string

I think I had tried this before and this was a success on gcc/Linux but I guess there is some issue with some other build. I don't really remember the issue. There was a discussion a few months ago in the IRC. I guess I'll have to dig that up :)
I am also not 100% sure if what you are thinking is what I am also thinking about (so please excuse me if you do the digging and find out that it was a different issue :) )

**Comment by prince776** [comment]
Thanks for all the information @sreenikSS 

> That is understandable since all possible layer types are supposed to be included here. However, what happens when you pass an instance of type `LayerTypes<>`?
> 
You mean when we  pass a layer which is not defined in LayerType. Then I don't understand the error, I get a huge error 
`, arma::Mat<double> >*, mlpack::ann::BaseLayer<mlpack::ann::SoftplusFunction, arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::BatchNorm<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::BilinearInterpolation<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Concat<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Concatenate<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::ConcatPerformance<mlpack::ann::NegativeLogLikelihood<arma::Mat<double>, arma::Mat<double> >, arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Constant<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Convolution<mlpack::ann::NaiveConvolution<mlpack::ann::ValidConvolution>, mlpack::ann::NaiveConvolution<mlpack::ann::FullConvolution>, mlpack::ann::NaiveConvolution<mlpack::ann::ValidConvolution>, arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::TransposedConvolution<mlpack::ann::NaiveConvolution<mlpack::ann::ValidConvolution>, mlpack::ann::NaiveConvolution<mlpack::ann::ValidConvolution>, mlpack::ann::NaiveConvolution<mlpack::ann::ValidConvolution>, arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::DropConnect<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Dropout<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::AlphaDropout<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::ELU<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::FlexibleReLU<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Glimpse<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::HardTanH<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Highway<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::Join<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::LayerNorm<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::LeakyReLU<arma::Mat<double>, arma::Mat<double> >*, mlpack::ann::CReLU<arma::Mat<double>, arma::Mat<double> >*,`  multiplied by 10.
It probably says that it's not defined within the boost::variant list `LayerTypes`.

> You mean to create a dummy layer, right? That should work too but try with `LayerTypes<>` once before doing that.
Yes It's possible to check if it's a custom layer with LayerTypes<> but I don't understand how can I retrieve the name of layer if that is not a member of class.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

**Pull request closed without merging**